### PR TITLE
104 display database stored post to the user

### PR DESCRIPTION
**Pull request by iliasmerigh**
### **In this PR I did the following :** 

- The trash photos taken by the logged-in user are now visible on his profile. This means that the posts belonging to him on screen are now synchronized with the database.
- Established the connection between the feed and the database. Currently, there is no intelligent algorithm in place to fetch and display relevant photos, so only the user's own photos are shown. In other words, the feed is linked to the database, but the implementation of the algorithm for selecting relevant photos will be carried out in the next sprint.
- The posts contain accurate information, including the user's name, description, and profile picture.
- Conducted testing on some significant components.

### **Issues encountered :**

I wasn't able to test everything, but the untested cases were already untouched prior to the synchronization with the database. These cases appear to be straightforward to test, yet inexplicably, they don't function as expected. I believe this is precisely why they were ignored in previous testing. For instance:
- Clicking the `profile_cam` button at the top left of the camera screen wasn't tested, and I couldn't successfully simulate the click in the tests.
- The hiding of an avatar in a post wasn't tested, and I also failed to create a successful test for this case.
- And so on...

To summarize, I attempted to write several missing tests that initially appeared easy, but I haven't been able to make them work, despite the correct behavior displayed on the screen. I suspect this is why these cases were previously disregarded.

Edit : Due to a strange behavior in the `UserPhotosViewAdapterTest`, some tests are non-deterministic. Sometimes they succeed and sometimes not. It especially fails when they is another test in the test file, even if this later one is completely independent from the first one. The expected behavior is working fine on screen and I managed to keep a good coverage. Please look at the comments in `UserPhotosViewAdapterTest` if you want more details.

### **Screenshots :**

Feed fragment and Profile fragment are sync with the database with old posts and new freshly added post :

image

image

### CODE REVIEW

File: app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt [hunk 1]
```diff
@@ -44,41 +58,43 @@ class UserPhotosViewAdapter(
...
-            holder.photoView.setImageBitmap(
-                BitmapFactory.decodeStream(
-                    url.openConnection().getInputStream()
-                )
-            )
-            //------------FOR DEMO -----------------
+                holder.avatarView.visibility = View.VISIBLE
+
+                // Display avatar if on feed
+                if (user?.profilePictureMetadata != null) {
```
Comment (david-schroeter) [inline app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt:73]: Here I would recommend you to write it like this. It prevents you from using the `!!` to assert a non-null value on the profilePictureMetadata.

```suggestion
                user?.profilePictureMetadata?.also {
                    val avatarImage = db.getUserProfilePicture(it, userId)
                    holder.avatarView.setImageBitmap(avatarImage)
                }
               
```

File: app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt [hunk 2]
```diff
@@ -44,41 +58,43 @@ class UserPhotosViewAdapter(
...
+
+                // Set default value
+                holder.titleView.text = user?.displayName ?: "Unknown user"
+                holder.subtitleView.text = (photo?.takenOn?.toString()
+                    ?: "Unknown date") + " | " + (photo?.trashCategory?.title
+                    ?: "No category")
+                holder.descriptionView.text = photo?.caption
+                holder.photoView.setImageBitmap(
+                    db.getImage(photo)
+                )
```
Comment (david-schroeter) [inline app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt:96]: Here it would be better if we had the strings in the String resource file. So that we can easily localize them.

There would be two option:

1. The best option to be able to achieve this, would be to separate the logic from the view, and delegate the logic to the caller of this Adapter, which in this case is `UserPhotoFragment`. So that we directly have the context. It would even allow you to call the `db` from outside the Adapter (Directly in the `UserPhotoFragment` class,) and therefor avoid to pass `lifecycleScope` as parameter to the adapter. You have an example of how I did it for the `EventsListAdapter` with the `EventDataAdapterListeners`. You would have to declare an interface, and then make the `UserPHotoFragment` implement this interface.
2. You can add the `context` as a parameter to the adapter, and directly access the string resources.

I personally find the first option cleaner, and it would also make it easier for you to test the Adapter I think, but the second option is probably easier.

Let me know what you think.

**Review by david-schroeter — CHANGES_REQUESTED** [review]
Thanks a lot for this new Features, it's nice to finally be able to see the photo that we take! Nice job!

I just add a few comments, let me know what you think about them.

I also saw that you added that for now in the feed, we only display our own pictures. I don't think it would be much harder to fetch all pictures and display them all together, if I am not mistaken Xavier already wrote something to do that in is draft PR about `Paging`. Let me know if you think it would be doable to still have that in this PR or rather not?

**Comment by 2Tricky4u** [comment]
Hey,
 I just checked the PR and I have some comments too =).
You mention profile_cam which isn't tested. It's the case with the test `fun clickOnProfileDetailsCameraRedirectCorrectly()` but in MainActivityTest for fragmentManager purposes.

Also as we mentioned orally during our meeting and also on what's app at the beginning and middle of the sprint, I highly suggested you to take a look at my Paging3 PR (and ask me questions about it if needed). The algorithm to sort the post by dates is totally done and also the fetching of the post of all following just needs a re-adaptation.

Doing so, would not make our feed dynamically fetch posts by batches but at least it would actually display actual posts. And the code is available on my PR.

Also If you still have a little bit of time, you could touch the XML files which was one of the first of the projects to better fit in the app UI as it would be some minor changes. (Like making the profile picture of a post rounded instead of rectangular, etc..)

Thank you for your current work and I can't wait to find out about our new feed =)

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/CameraWithUIFragmentTest.kt [hunk 3]
```diff
@@ -29,6 +29,7 @@ import org.junit.runner.RunWith
 import org.koin.dsl.module
 import org.mockito.Mockito
 import org.mockito.Mockito.`when`
+import org.mockito.kotlin.whenever
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/CameraWithUIFragmentTest.kt:32]: Well spotted !

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/ProfileDetailsFragmentTest.kt [hunk 4]
```diff
@@ -47,22 +47,24 @@ import java.util.*
...
+                userId1
             ),
 
             ProfilePhotoMetadata(
                 "erfs",
                 ParcelableDate.now,
-                "0"
+                userId1
             )
         )
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/ProfileDetailsFragmentTest.kt:65]: You are right, I did this mistake in a previous PR and I didn't pay attention to change it. Thank you I made the change.

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/activities/MainActivityTest.kt [hunk 5]
```diff
@@ -175,19 +178,22 @@ class MainActivityTest {
...
 
-
     @Test
     fun pressFeedMenuDisplayFeedFragment() {
-        onView(withId(R.id.bottomMenuFeed))
-            .check(matches(isDisplayed()))
-            .perform(click())
 
-        onView(withId(R.id.feed_list)).check(matches(isDisplayed()))
+        runTest {
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/activities/MainActivityTest.kt:184]: I was not really familiar with the `runTest` this is why I put it. I thought that we have to use the `runTest` when a called method uses `lifecyleScope.launch`, this is why I put it. I've just removed it and it works, thank you

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 6]
```diff
@@ -1,32 +1,110 @@
...
 import org.hamcrest.CoreMatchers
 import org.hamcrest.CoreMatchers.equalTo
 import org.hamcrest.MatcherAssert
+import org.junit.BeforeClass
 import org.junit.Rule
 import org.junit.Test
+import org.junit.runner.RunWith
+import org.koin.dsl.module
+import org.mockito.Mockito
+import org.mockito.kotlin.whenever
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:37]: I forgot to remove them after trying to debug. I've removed them !

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 7]
```diff
@@ -1,32 +1,110 @@
...
 
+@RunWith(AndroidJUnit4::class)
+@LargeTest
 class UserPhotosViewAdapterTest {
+
+    companion object {
+        private val db: DB = Mockito.mock(DB::class.java)
+        private val auth: Auth = Mockito.mock(Auth::class.java)
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:45]: I tried to do the same as other test classes but I'm not sure to understand what you mean by :

> import .mock statically

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 8]
```diff
@@ -1,32 +1,110 @@
...
+        fun setUp() {
+            // The implementation need to be provided before the rule is executed,
+            // that's why we do it in the beforeClass method
+            runTest {
+                // setup basic get user and getProfilePicture use in multiple tests
+                Mockito.`when`(db.getUser(user.id))
+                    .thenReturn(user)
+                // add a small delay, just to be sure that it is triggered after initialization
+                // and arrive second, after the initial null value
+                // user between tests, by simply pushing a new userId
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:75]: Yes you are right. I will also remove it from `CameraUIWithFragment` (I took some time to understand that it was also unrelated with this class but only in the `MainActivityTest`)

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 9]
```diff
@@ -1,32 +1,110 @@
...
+
+        @OptIn(ExperimentalCoroutinesApi::class)
+        @BeforeClass
+        @JvmStatic
+        fun setUp() {
+            // The implementation need to be provided before the rule is executed,
+            // that's why we do it in the beforeClass method
+            runTest {
+                // setup basic get user and getProfilePicture use in multiple tests
+                Mockito.`when`(db.getUser(user.id))
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:71]: As I said previouly, I don't undersand what you mean by import it statically since I think I did the same as other test classes. Could you enlighten me a bit please ?

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 10]
```diff
@@ -1,32 +1,110 @@
...
+            User("1", 123, "Alice"),
+            User("2", 0, "Bob Zeu bricoleur"),
+            User("3", 14, "Charlie Chaplin"),
+            User("4", 23, "David Pujadas"),
+            User("5", 10492, "Euler"),
+            User("6", 1234, "Alain Berset"),
+            User("7", 1235, "Mister Alix")
+        )
+        private val fakePicture = Bitmap.createBitmap(120, 120, Bitmap.Config.ARGB_8888)
+        private val authUserFlow = MutableStateFlow(userId)
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:60]: Well seen !

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 11]
```diff
@@ -1,32 +1,110 @@
...
+   @get:Rule
+    val koinTestRule = KoinTestRule(
+        modules = listOf(module {
+            single { db }
+            single { auth }
+        })
+    )
+
+    //@get:Rule
+    //val koinTestRule = KoinTestRule()
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:101]: Yes I forgot to remove it after debugging. Thank you

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (2Tricky4u) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:45]: I think David means that you use: `import org.mockito.Mockito.mock` in imports and so you just need to write mock instead of Mockito.mock

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 12]
```diff
@@ -36,25 +114,66 @@ class UserPhotosViewAdapterTest {
 
     @Test
     fun userViewAdapterGetItemCountWorksOnEmptyList() {
-        userPhotoViewAdapter = UserPhotosViewAdapter(listOf(), true)
+
+        val userPhotoViewAdapter = UserPhotosViewAdapter(listOf(), true, TestLifecycleOwner().lifecycleScope)
         MatcherAssert.assertThat(userPhotoViewAdapter.itemCount, equalTo(0))
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:119]: I was already like that in other test methods so I thought about changing it. But you are right, I made the change

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:71]: I made the change

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:45]: Good practice ! I made the change

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 13]
```diff
@@ -36,25 +114,66 @@ class UserPhotosViewAdapterTest {
...
+        runTest {
+
+            Mockito.`when`(db.getUser(user.id))
+                .thenReturn(null)
+
+            val viewHolder = userPhotoViewAdapter.onCreateViewHolder(LinearLayout(appContext), 0)
+
+            MatcherAssert.assertThat(viewHolder.titleView.text, equalTo("Title"))
+            MatcherAssert.assertThat(viewHolder.subtitleView, CoreMatchers.notNullValue())
+            MatcherAssert.assertThat(viewHolder.subtitleView.text.toString(), equalTo("subhead"))
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:141]: I noticed it in the previous version of the test and I had a similar observation as you. However, considering that it was included in the code, I assumed there might have been a reason for it. Now that we both agree it doesn't make sense, I will remove it, thank you

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 14]
```diff
@@ -36,25 +114,66 @@ class UserPhotosViewAdapterTest {
...
+            MatcherAssert.assertThat(viewHolder.avatarView.visibility, equalTo(View.VISIBLE))
+
+            /*
+            MatcherAssert.assertThat(viewHolder.titleView.text, equalTo(user.displayName))
+            MatcherAssert.assertThat(viewHolder.subtitleView, CoreMatchers.notNullValue())
+            MatcherAssert.assertThat(viewHolder.subtitleView.text.toString(), CoreMatchers.containsString(TrashCategory.PLASTIC.title))
+            MatcherAssert.assertThat(
+                viewHolder.descriptionView.text.toString(),
+                equalTo("Look at me cleaning!")
+            )*/
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:175]: I ran into a situation where I couldn't test it for some reason, even though the expected behavior was showing up correctly on the screen. I suspect there might be a slight delay before the views update due to data fetching, but I am not sure if it's the actual reason. It would have been really nice to test this case, but unfortunately, I wasn't able to figure it out. If anyone has any ideas or suggestions on how to tackle this, I would really appreciate your help !

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/DB.kt [hunk 15]
```diff
@@ -117,7 +116,6 @@ interface DB {
     /**
      * Retrieves the profile image associated with the given [userId] and [metadata] from the database
      *
-     * @param metadata the metadata associated with the given image we want to retrieve
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/DB.kt:120]: Yes I removed it first and forgot to put it back. Well spotted !

File: app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt (hunk 1 above)
Comment (iliasmerigh) [reply app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt:73]: You are right, it cleaner ! Thank you

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (2Tricky4u) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt:175]: If it is about a delay for the fragment to show you could try to use BaseRobot in the test folder EspressoUtils. 
It is used in the following manner:
`// Click on element withId
BaseRobot().doOnView(withId(R.id.viewIWantToFind), click())

// Assert element withId is displayed
BaseRobot().assertOnView(withId(R.id.viewIWantToFind), matches(isDisplayed()))`

Maybe you won't be able to test as much as you would like but it should deal with the delay if it's the case

File: app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt (hunk 2 above)
Comment (iliasmerigh) [reply app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt:96]: Actually, it was already like that in the previous version. I tried to link it with the ressource file but I didn't manage to do it and I think this is the reason why it was left like that. I will try to implement what you suggested. Thank you very much.

Comment (iliasmerigh) [reply app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt:96]: I took inspiration of what was already implemented in `UserViewAdapter` which correspond to the option 2. Thank you very much for your help

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt [hunk 16]
```diff
@@ -102,68 +110,157 @@ class UserPhotosViewAdapterTest {
...
+        the added test is completely unrelated with this one. I tried to declare all variables inside this method
+        in order to avoid to modify a shared state variable but it still doesn't work (including the ressource and appContext).
+        I tried to launch and close the activity inside this method but it's still not working.
+        When I comment back other tests, sometimes this one fails and sometimes it passes.
+        In other words, the way this test file is written makes this test non-determinisic.
+        But I preferred to keep these test so we ensure that the app doesn't crash because by trying several paths. Actually,
+        these tests helped me to debug by crashing the app when I made a mistake, so I think that they are relevant, and we could
+        improve them in the future by finding were the problem comes from (this is why I left some unused variable such as dateString).
+        It also increases the coverage by testing some
+        non-trivial path to check that the app doesn't fail.
```
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt]: I changed them, and I tried to minimize the "we" as well by turning them into a passive form when it was possible

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (iliasmerigh) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapterTest.kt]: Thank you for your explanations, I tried to implement was you said but it seems that the problem is elsewhere. I modified the tests and explained everything in the comments. Thanks anyway for your help :)

**Comment by iliasmerigh** [comment]
1)

> You mention profile_cam which isn't tested. It's the case with the test fun clickOnProfileDetailsCameraRedirectCorrectly() but in MainActivityTest for fragmentManager purposes.

I apologize for not noticing that earlier. Thank you for pointing it out.

2)

Thank you @2Tricky4u and @david-schroeter so much for your explanations. I really appreciate your help and explanations with **feed3**.

The task clearly states that we need to **"_display database stored posts to the user._"**. When I noticed that there was no algorithm in the code I had planned to start implementing it, but I realized I was deviating from the task, which could have caused huge problems like in the previous sprint. That's why I decided to leave it for a separate task. It actually makes sense because the criteria for the fetching algorithm are independent of the database and the tests. As @LucaCarroz  said, It's not a matter of difficulty, but I learned from a mistake in the last sprint and wanted to avoid falling into the same trap: splitting the work of each task to make it easier to debug and review.

I managed to complete this task as I defined it, and I believe it's not a good choice to add the algorithm at the end. Moreover, its inclusion is not necessary for the correctness of the current implementation. Even though it's already implemented in _feed3_, understanding how it works will take some time, and honestly, it won't be ready before a while. **Additionally, when it is eventually reviewed, it will be considered independently of what has already been implemented.**

In conclusion, I hope my explanation helps you understand why I believe it would be best to handle the algorithm in a separate task. Alternatively, if @2Tricky4u is interested in working on it since he already has some experience with it, he can let me know if he prefers to further develop his algorithm. Otherwise, I don't have any problem to work on it for my next task.

**Comment by 2Tricky4u** [comment]
> The task clearly states that we need to **"_display database stored posts to the user._"**

We also stated multiple times what was the point of your task and even after reading "display database stored posts to the user" I still understand that the posts in the databases are fetched by the current user of all his following. Like in all social networks of the last decades. Or maybe I'm totally hallucinating for the last decades and I'm just seeing my own post 
3e3ce20d7980b7e336a9b3d4bd9f0343

**Comment by 2Tricky4u** [comment]
Fetch followings post to adapt:
```
private val db  by inject<DB>(FirebaseDB::class.java)

    /** Function to retrieve the feed of the current user */
    fun fetchFeed(): Flow<PagingData<PhotoMetadata>> {
        var feedPosts: List<PhotoMetadata> = listOf()
        runBlocking {
            viewModelScope.launch {
                feedPosts = currentUser.value?.let { cur_user ->
                    cur_user.following?.flatMap { id ->
                        getUserPosts(id) ?: emptyList()
                    }
                } ?: emptyList()
            }
        }
        //sort it by date
        return repository.letPhotoMetadataFlow(feedPosts.sortedBy { it.takenOn})
    }


    suspend private fun getUserPosts(user : String) : List<PhotoMetadata> {
        return db.getUser(user).let { us ->
            us?.posts ?: emptyList()
        }
    }
}
```

To add to ParcelabaleDate:
```
 class ParcelableDate() : Parcelable, Comparable<ParcelableDate> {
            ...
 override fun compareTo(other: ParcelableDate): Int {
        return date?.compareTo(other.date) ?: 0
    }
```

I think it's the main portion of code you can reuse (I'm not sure if there is more or not). Could you try to use it very rapidly and if it doesn't work, no worries we do it next sprint?

**Review by LucaCarroz — APPROVED** [review]

File: app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt [hunk 17]
```diff
@@ -42,43 +47,43 @@ class UserPhotosViewAdapter(
...
+            }
+
+            // Display post content
+            val dateString = (photo.takenOn?.toString() ?: resources.getString(R.string.unknown_date))
+            val categoryString = (photo.trashCategory?.title ?: resources.getString(R.string.no_category))
 
-            // TODO : to change
-            holder.descriptionView.text = photo?.caption
+            holder.titleView.text = user.displayName ?: resources.getString(R.string.unknown_user)
+            holder.subtitleView.text = "$dateString | $categoryString"
```
Comment (david-schroeter) [inline app/src/main/java/com/github/sdp_begreen/begreen/fragments/UserPhotosViewAdapter.kt:84]: Here I would use a resource string with format. See this [link](https://stackoverflow.com/questions/12627457/format-statement-in-a-string-resource-file) for an example

**Review by david-schroeter — CHANGES_REQUESTED** [review]
Most of the changes looks good to me, there is just a small improvement that you can do regarding the way you use the string resources. I added a link where you can see an example

**Review by david-schroeter — APPROVED** [review]
The changes you made look all good to me, thanks for the implementation of these features.

However, I do share the point of view of Xavier that we discussed a lot about what we expected from this task, and our expectations were not fulfilled. 

We will know that for next time, we should precisely write down on the task what we expect from it, so that we all agree on what should be implemented.

Anyway, thank again for your work it's already a great improvement, and I can't wait to see it fully functional in the following sprint.
